### PR TITLE
AP_Scripting: save some flash

### PR DESCRIPTION
**Pull request by IamPete1**
This allows initialization of unit32 from negative numbers, that allows us to pass negative inits that can be represented by lua ints instead of large positive numbers that loose precision as they are represented as lua floats.

Also saves some flash by re-using the div function for idiv and reusing the auto generated binding arg checker.

### CODE REVIEW

**Comment by WickedShell** [comment]
The motivator appears to be some oddities with using literals in scripts, aggravated by rejecting negative integer values. Given the need to have useful literals it might be better to just remove this negative check for uint32_t boxed types.

**Comment by IamPete1** [comment]
Example test script: 

```
  local f = uint32_t(4157483647)
  gcs:send_text(0, string.format("uint32_t(4157483647):toint() = %i", f:toint()) .. " string: " .. tostring(f))

  local f = uint32_t(10472251) * uint32_t(397)
  gcs:send_text(0, string.format("(uint32_t(10472251) * uint32_t(397)):toint() = %i", f:toint()) .. " string: " .. tostring(f))


  local f = uint32_t(-137483649)
  gcs:send_text(0, string.format("uint32_t(-137483649):toint() = %i", f:toint()) .. " string: " .. tostring(f))
```

File: libraries/AP_Scripting/lua_boxed_numerics.cpp [hunk 1]
```diff
@@ -20,14 +20,14 @@ uint32_t coerce_to_uint32_t(lua_State *L, int arg) {
 
         int success;
         const lua_Integer v = lua_tointegerx(L, arg, &success);
-        if (success && v >= 0) {
+        if (success && v >= INT32_MIN) {
```
Comment (WickedShell) [inline libraries/AP_Scripting/lua_boxed_numerics.cpp:23]:
```suggestion
        if (success) {
```

I'm pretty sure you are accidentally forcing a number of things into the float path at the moment? I was suggesting as long as we were passed an integer we just cast it, what you have here is ignoring many integer cases....

File: libraries/AP_Scripting/lua_boxed_numerics.cpp [hunk 2]
```diff
@@ -20,14 +20,14 @@ uint32_t coerce_to_uint32_t(lua_State *L, int arg) {
...
             return static_cast<uint32_t>(v);
         }
     }
     { // float
         int success;
         const lua_Number v = lua_tonumberx(L, arg, &success);
-        if (success && v >= 0 && v <= float(UINT32_MAX)) {
+        if (success && v >= float(INT32_MIN) && v <= float(UINT32_MAX)) {
```
Comment (WickedShell) [inline libraries/AP_Scripting/lua_boxed_numerics.cpp:30]: Same question as above.

**Review by WickedShell — CHANGES_REQUESTED** [review]

File: libraries/AP_Scripting/generator/src/main.c [hunk 3]
```diff
@@ -2339,7 +2339,7 @@ void emit_sandbox(void) {
 void emit_argcheck_helper(void) {
   // tagging this with NOINLINE can save a large amount of flash
   // but until we need it we will allow the compilier to choose to inline this for us
-  fprintf(source, "static int binding_argcheck(lua_State *L, int expected_arg_count) {\n");
+  fprintf(source, "int binding_argcheck(lua_State *L, int expected_arg_count) {\n");
```
[the remark this answers is not in the document]
Comment (IamPete1) [reply libraries/AP_Scripting/generator/src/main.c:2342]: To get it to build without lots of fiddling. Any reason it needs to be static?

**Comment by IamPete1** [comment]
```
Binary Name      Text [B]         Data [B]     BSS (B)        Total Flash Change [B] (%)      Flash Free After PR (B)
---------------  ---------------  -----------  -------------  ----------------------------  -------------------------
antennatracker   -988 (-0.0729%)  0 (0.0000%)  -4 (-0.0015%)  -988 (-0.0728%)                                  610524
arducopter-heli  -988 (-0.0553%)  0 (0.0000%)  -4 (-0.0015%)  -988 (-0.0552%)                                  175748
ardurover        -988 (-0.0615%)  0 (0.0000%)  4 (+0.0015%)   -988 (-0.0615%)                                  359660
arducopter       -988 (-0.0555%)  0 (0.0000%)  4 (+0.0015%)   -988 (-0.0554%)                                  182956
ardusub          -988 (-0.0631%)  0 (0.0000%)  -4 (-0.0015%)  -988 (-0.0630%)                                  398760
blimp            -988 (-0.0762%)  0 (0.0000%)  4 (+0.0015%)   -988 (-0.0761%)                                  668376
arduplane        -988 (-0.0560%)  0 (0.0000%)  4 (+0.0015%)   -988 (-0.0559%)                                  200196
```

File: libraries/AP_Scripting/generator/src/main.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (IamPete1) [reply libraries/AP_Scripting/generator/src/main.c:2342]: I would argue that none of the other functions we define in the header are static, that is the new and check functions for userdata and the load functions.

**Comment by tridge** [comment]
local foo = uint32(0xF000,0x7)

**Comment by IamPete1** [comment]
Dropped controversial signed vs unsigned thing, now just the flash savings.

I will investigate the cost of moving to doubles in lua vs keeping uint32.

**Comment by WickedShell** [comment]
> we could check if the constructor has 2 arguments we can do: local foo = uint32_t(0xF000, 0x0007) that would create 0xF0000007

Why not accept 0xFFFFFFFF directly? The compiler will turn it into -1 but that will be an integer literal? That's how normal lua code works at that point, and is far clearer...

**Review by tridge — APPROVED** [review]

**Pull request merged**